kernel/fatfs.c

```diff
@@ -133,10 +134,47 @@ STATIC void init_direntry(struct dirent *dentry, unsigned attrib,
 int dos_open(char *path, unsigned flags, unsigned attrib, int fd)
 {
   REG f_node_ptr fnp = sft_to_fnode(fd);
-  int status = find_fname(path, D_ALL | attrib, fnp);
+  int status;
+
+  /* Special handling for volume name (they can coexist with files/directories) */
+  if ((attrib & D_VOLID) &&
+      ((attrib & D_DIR) != D_DIR) && ((attrib & D_LFN) != D_LFN) &&
+      (flags & O_CREAT) && (flags & O_FCB))
+  {
+    int ret;
+    char rpath[16];
+    char *p;
+
+    /* Check to see if there are any other files in the root with VOLID set */
+    rpath[0] = path[0];
+    rpath[1] = ':';
+    rpath[2] = '\\';
+    rpath[3] = '\0';
+    fnp = dir_open(rpath, FALSE, fnp);
+    status = find_in_dir(D_VOLID, D_VOLID | D_RDONLY | D_ARCHIVE, fnp);
+    if (status == SUCCESS)
+      return DE_ACCESS;
+
+    /* Find the filename.ext component of the path. Note that MS-DOS 6.22 and
+     * DR-DOS 7.01 set the volume label (in the root) when a caller tries to
+     * create a file with VOLID set in a non root directory */
+    p = strrchr(path, '\\');
+    if (!p)
+      return DE_ACCESS;
+    strncpy(rpath + 3, p + 1, sizeof(rpath) - 3);
+
+    /* Create our new label */
+    ret = alloc_find_free(fnp, rpath);
+    if (ret != SUCCESS)
+      return ret;
+    status = S_CREATED;
+
+    goto doit;
```

Comment:
Maybe use `else` here?
You don't even need to re-indent
everything below.
You can instead do
`if (find_fname(path, D_ALL | attrib, fnp) == SUCCESS)`
and get the same indentation.

Reply:
Why?

You currently have:
```
   goto doit;
  }
  status = find_fname(path, D_ALL | attrib, fnp);
  if (status == SUCCESS)
  {
  }
doit:
```
If you put `find_fname()` directly into the `if`,
then you can put `else` and not re-indent?
```
  }
  else if (find_fname(path, D_ALL | attrib, fnp) == SUCCESS)
  {
  }
doint: //unneeded now
```

Reply:
> 
You currently have:
~~~
   goto doit;
  }
  status = find_fname(path, D_ALL | attrib, fnp);
  if (status == SUCCESS)
  {
  }
doit:
~~~

No quite, there's an 'else if' and an else following
```c
    goto doit;
  }                                                                             
                                                                                
  /* Check that we don't have a duplicate name, so if we  */                    
  /* find one, truncate it (O_CREAT).                     */                    
  status = find_fname(path, D_ALL | attrib, fnp);                               
  if (status == SUCCESS)                                                        
  {                                                                             
    <snip>                                                                      
  }                                                                             
  else if (status == DE_FILENOTFND && (flags & O_CREAT))                        
  {                                                                             
    int ret = alloc_find_free(fnp, path);                                       
    if (ret != SUCCESS)                                                         
      return ret;                                                               
    status = S_CREATED;                                                         
  }                                                                             
  else                                                                          
  {                                                                             
    /* open: If we can't find the file, just return a not    */                 
    /* found error.                                          */                 
    return status;                                                              
  }                                                                             
                                                                                
doit:                                                                           
                                                                                
```

BTW I hate this style of indention, but there's no changing it now I suppose.

Reply:
Ah, I see, I messed up with that indentation. :)
Well, there is still an option of using
`switch (find_fname(path, D_ALL | attrib, fnp))`
and not indent `case`s, but in that case you'll
have to mess a bit with the O_CREAT part.

So whatever you like.
So should this be applied?

Reply:
Yes, I think it's good to apply as is.